clang/lib/CConv/DeclRewriter.cpp

```diff
@@ -0,0 +1,410 @@
+
+    if (Verbose) {
+      errs() << "Replacing type of decl:\n";
+      N.Declaration->dump();
+      errs() << "with " << N.Replacement << "\n";
+    }
+
+    // Get a FullSourceLoc for the start location and add it to the
+    // list of file ID's we've touched.
+    SourceRange tTR = N.Declaration->getSourceRange();
+    FullSourceLoc tFSL(tTR.getBegin(), A.getSourceManager());
+    TouchedFiles.insert(tFSL.getFileID());
+
+    // Exact rewriting procedure depends on declaration type
+    if (N.hasDeclType<ParmVarDecl>()) {
+      // TODO: why is this asserted?
+      assert(N.Statement == nullptr);
+      rewriteParmVarDecl(N);
+    } else if (N.hasDeclType<VarDecl>()) {
+      rewriteVarDecl(N, ToRewrite);
+    } else if (N.hasDeclType<FunctionDecl>()) {
+      rewriteFunctionDecl(N);
+    } else if (N.hasDeclType<FieldDecl>()) {
+      SourceRange SR = N.getDecl<FieldDecl>()->getSourceRange();
+      if (canRewrite(R, SR))
+        R.ReplaceText(SR, N.Replacement);
+    }
+  }
+}
+
+std::map<std::string, std::string> DeclRewriter::NewFuncSig;
+
+void DeclRewriter::rewriteDecls(ASTContext &Context, ProgramInfo &Info,
+                                Rewriter &R, std::set<FileID> &TouchedFiles) {
+  // Compute the bounds information for all the array variables.
+  ArrayBoundsRewriter ABRewriter(&Context, Info);
+
+  // Collect function and record declarations that need to be rewritten in a set
+  // as well as their rewriten types in a map.
+  RSet RewriteThese(DComp(Context.getSourceManager()));
+  TypeRewritingVisitor TRV = TypeRewritingVisitor(&Context, Info, RewriteThese,
+                                                  NewFuncSig, ABRewriter);
+  StructVariableInitializer SVI = StructVariableInitializer(&Context, Info,
+                                                            RewriteThese);
+  for (const auto &D : Context.getTranslationUnitDecl()->decls()) {
+    TRV.TraverseDecl(D);
+    SVI.TraverseDecl(D);
+  }
+
+  // Build a map of all of the PersistentSourceLoc's back to some kind of
+  // Stmt, Decl, or Type.
+  TranslationUnitDecl *TUD = Context.getTranslationUnitDecl();
+  std::set<PersistentSourceLoc> Keys;
+  for (const auto &I : Info.getVarMap())
+    Keys.insert(I.first);
+  MappingVisitor MV(Keys, Context);
+  for (const auto &D : TUD->decls())
+    MV.TraverseDecl(D);
+  SourceToDeclMapType PSLMap;
+  VariableDecltoStmtMap VDLToStmtMap;
+  std::tie(PSLMap, VDLToStmtMap) = MV.getResults();
+
+  // Add declarations from this map into the rewriting set
+  for (const auto &V : Info.getVarMap()) {
+    PersistentSourceLoc PLoc = V.first;
+    CVarSet Vars = V.second;
+    // I don't think it's important that Vars have any especial size, but
+    // at one point I did so I'm keeping this comment here. It's possible
+    // that what we really need to do is to ensure that when we work with
+    // either PV or FV below, that they are the LUB of what is in Vars.
+    // assert(Vars.size() > 0 && Vars.size() <= 2);
```

Comment:
I think this is stale. I believe it's a remnant of a more brittle time when we didn't handle prototypes and defs uniformly, when rewriting.